examples/DDS: Cleanup, remove sleep, etc.

ChristianTackeGSI opened the pull request:
We should get the DDS tests to the point, that they do not depend on "sleep" to synchronize state. So let's go the other way: Remove the sleeps to trigger them and let's fix them!

To make the output more readable, include some blank lines and reformat headers in a more consistant way.

---

Checklist:

* [X] Rebased against `dev` branch
* [X] My name is in the resp. CONTRIBUTORS/AUTHORS file
* [X] Followed [the seven rules of great commit messages](https://chris.beams.io/posts/git-commit/#seven-rules)

dennisklein commented:
I do not recommend adding sleeps at all. Why are they needed at all (the ones that are already present and are prolonged in this PR)?

rbx commented:
I agree with Dennis. Remove all the sleeps. If then something doesn't work, open an issue in FairMQ SDK/fairmq-dds-command-ui/DDS.

dennisklein commented:
Please note, that the CLI usage of `fairmq-dds-command-ui` has changed over time, see our canonical example for reference:
* FairMQ `v1.4.[8-12]`: https://github.com/FairRootGroup/FairMQ/blob/v1.4.8/examples/dds/fairmq-start-ex-dds.sh.in
* FairMQ `v1.4.13+`: https://github.com/FairRootGroup/FairMQ/blob/v1.4.13/examples/dds/fairmq-start-ex-dds.sh.in

dennisklein commented:
Also note: https://github.com/FairRootGroup/FairMQ/issues/303

ChristianTackeGSI commented:
The major issue is with the jun19 environment in FairSoft.
It has an old fairmq and dds version.
So I assume, that some races/etc have already been fixed in newer versions.

Hmmmm…

The alternative option is to ignore those tests in the jun19 case.
I'll look into this and into some other details.

ChristianTackeGSI commented:
Thanks for the heads up!

I have rewritten this PR to go the other way: Let's find all the races!

So I have removed most of the sleeps. Should we remove the `sleep 2` also?

And to be able to read the output more easily, I have added a bit of empty lines, etc.

ChristianTackeGSI commented:
Okay, dropping racy tests on FairSoft testsuite for jun19: https://github.com/FairRootGroup/FairSoft/pull/343

dennisklein commented on examples/MQ/pixelDetector/run/scripts/test-pixelSim.sh.in:
```diff
@@ -51,37 +51,44 @@ dds-topology -s $DDS_SESSION_ID --activate $FAIRMQ_DDS_TOPO_FILE --disable-valid
-sleep 1
+for cmd in s t d q
+do
+	echo ""
+	echo "***** fairmq-dds-command-ui -c $cmd"
```
Isn't `set -x` doing what you are you doing here?

ChristianTackeGSI replied:
Yes, `set -x` would do that. But do we want the whole other test logic to spam the logs?
What about rephrasing the output? Like "Sending command $cmd"?

dennisklein replied:
```
set -x
fairmq-dds-command-ui -c s
fairmq-dds-command-ui -c t
fairmq-dds-command-ui -c d
fairmq-dds-command-ui -c q
set +x
```
What about this?

I am honestly not sure if the loop is increasing readability here.

dennisklein commented on examples/MQ/pixelDetector/run/scripts/test-pixelSim.sh.in:
```diff
@@ -51,37 +51,44 @@ dds-topology -s $DDS_SESSION_ID --activate $FAIRMQ_DDS_TOPO_FILE --disable-valid
-fairmq-dds-command-ui -c x
-fairmq-dds-command-ui -c j
-fairmq-dds-command-ui -c r
+for cmd in IDLE i k b x j r
+do
+	echo ""
+	if (( ${#cmd} == 1 ))
+	then
+		echo "***** fairmq-dds-command-ui -c $cmd"
+		fairmq-dds-command-ui -c "$cmd"
+	else
+		echo "***** Waiting for state $cmd"
+		time fairmq-dds-command-ui -w "$cmd"
+	fi
+done
```
In FairMQ < `1.4.13` this needs to be

```suggestion
fairmq-dds-command-ui -w "IDLE" -n ${requiredNofAgents}
fairmq-dds-command-ui -c i -w "INITIALIZING DEVICE" -n ${requiredNofAgents}
fairmq-dds-command-ui -c k -w "INITIALIZED" -n ${requiredNofAgents}
fairmq-dds-command-ui -c b -w "BOUND" -n ${requiredNofAgents}
fairmq-dds-command-ui -c x -w "DEVICE READY" -n ${requiredNofAgents}
fairmq-dds-command-ui -c j -w "READY" -n ${requiredNofAgents}
fairmq-dds-command-ui -c r
```

The `-n` argument expects the number of device addressed.

dennisklein replied:
See https://github.com/FairRootGroup/FairMQ/blob/93eb599df7d9ee4c869e2fe70c03b7a845739683/examples/dds/fairmq-start-ex-dds.sh.in#L57-L72

dennisklein commented:
> So I have removed most of the sleeps. Should we remove the sleep 2 also?

This works only in DDS 2.4+
https://github.com/FairRootGroup/FairMQ/blob/93eb599df7d9ee4c869e2fe70c03b7a845739683/examples/dds/fairmq-start-ex-dds.sh.in#L33-L40

The commented lines work only in DDS 3.0+
https://github.com/FairRootGroup/FairMQ/blob/93eb599df7d9ee4c869e2fe70c03b7a845739683/examples/dds/fairmq-start-ex-dds.sh.in#L47-L52

dennisklein commented:
I would just disable this example if there is no DDS `3.0+` and no FairMQ `1.4.13+` - otherwise you will have to write multiple versions of the script or clutter everything up with unreadable if-else cases which defeats the purpose as an example.

FairMQ 1.4.3 and DDS 2.2 are just too old and cumbersome.

ChristianTackeGSI commented:
> I would just disable this example if there is no DDS `3.0+` and no FairMQ `1.4.13+` - otherwise you will have to write multiple versions of the script or clutter everything up with unreadable if-else cases which defeats the purpose as an example.

That sounds like a very good idea. If that's okay for everybody?

dennisklein commented:
Once #1020 is green, let's (merge it and) rebase this PR on top. It should repair the macOS CI builds.

dennisklein commented:
Please rebase on current `dev`

ChristianTackeGSI commented:
Okay, I think, I'm done for now.
It's not the full blown everything upgraded PR, but the "one step after the other" one.

karabowi commented:
I agree, I tried to write two different versions of the script, one for old DDS and FairMQ, and another one for the new versions.
It would be much cleaner to simply focus on the new one.
